{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Final Project of Introduction to Bioinformatics\n",
    "\n",
    "## Comparative Mitochondrial Genomics and Phylogenetic Analysis\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exploration and Reflection\n",
    "\n",
    "As we proceed with our analysis of mitochondrial DNA for phylogenetic tree construction, it is valuable to contemplate a few questions. These inquiries aim to facilitate a more thorough understanding of the roles and characteristics of mitochondrial DNA in the context of evolutionary biology:\n",
    "\n",
    "1. **Maternal Inheritance and Its Implications**: How does the maternal inheritance of mitochondrial DNA simplify our understanding of evolutionary lineage compared to nuclear DNA, which undergoes recombination? What unique insights can this aspect provide in tracing the evolutionary history of species?\n",
    "\n",
    "2. **Mutation Rate and Evolutionary Insights**: Mitochondrial DNA mutates at a faster rate than nuclear DNA. How does this characteristic make mtDNA a more sensitive tool for detecting recent evolutionary events and relationships among closely related species? Can you think of any specific scenarios or studies where this property of mtDNA has been particularly instrumental?\n",
    "\n",
    "Reflect on these questions as you work through the project, and consider how the properties of mitochondrial DNA enhance its value and applicability in evolutionary biology and beyond. Provide your answer either in this notebook, or in your report (if you had one).\n",
    "\n",
    "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
    "    <strong>Don't forget to answer these questions!</strong>\n",
    "</blockquote>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 0: Installing Necessary Packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "biopython (1.83) is installed\n",
      "pandas (2.1.4) is installed\n",
      "numpy (1.26.1) is installed\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import subprocess\n",
    "import pkg_resources\n",
    "\n",
    "def install(package):\n",
    "    subprocess.check_call([sys.executable, \"-m\", \"pip\", \"install\", package])\n",
    "\n",
    "REQUIRED_PACKAGES = [\n",
    "    'biopython',\n",
    "    'pandas',\n",
    "    'numpy'\n",
    "]\n",
    "\n",
    "for package in REQUIRED_PACKAGES:\n",
    "    try:\n",
    "        dist = pkg_resources.get_distribution(package)\n",
    "        print('{} ({}) is installed'.format(dist.key, dist.version))\n",
    "    except pkg_resources.DistributionNotFound:\n",
    "        print('{} is NOT installed'.format(package))\n",
    "        install(package)\n",
    "        print('{} was successfully installed.'.format(package))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import necessary libraries. \n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>taxo_id</th>\n",
       "      <th>specie</th>\n",
       "      <th>blast_name</th>\n",
       "      <th>genbank_common_name</th>\n",
       "      <th>accession_number</th>\n",
       "      <th>mtDNA</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8945.0</td>\n",
       "      <td>Eudynamys scolopaceus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Asian koel</td>\n",
       "      <td>NC_060520</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>7460.0</td>\n",
       "      <td>Apis mellifera</td>\n",
       "      <td>Bees</td>\n",
       "      <td>honey bee</td>\n",
       "      <td>NC_051932</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>36300.0</td>\n",
       "      <td>Pelecanus crispus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Dalmatian pelican</td>\n",
       "      <td>OR620163</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10116.0</td>\n",
       "      <td>Rattus norvegicus</td>\n",
       "      <td>Rodents</td>\n",
       "      <td>Norway rat</td>\n",
       "      <td>NC_001665</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_001665</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9031.0</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>birds</td>\n",
       "      <td>Gallus gallus</td>\n",
       "      <td>NC_053523</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   taxo_id                 specie blast_name genbank_common_name  \\\n",
       "0   8945.0  Eudynamys scolopaceus      birds          Asian koel   \n",
       "1   7460.0         Apis mellifera       Bees           honey bee   \n",
       "2  36300.0      Pelecanus crispus      birds   Dalmatian pelican   \n",
       "3  10116.0      Rattus norvegicus    Rodents          Norway rat   \n",
       "4   9031.0          Gallus gallus      birds       Gallus gallus   \n",
       "\n",
       "  accession_number                                              mtDNA  \\\n",
       "0        NC_060520  https://www.ncbi.nlm.nih.gov/nucleotide/NC_060...   \n",
       "1        NC_051932  https://www.ncbi.nlm.nih.gov/nucleotide/NC_051...   \n",
       "2         OR620163    https://www.ncbi.nlm.nih.gov/nuccore/OR620163.1   \n",
       "3        NC_001665     https://www.ncbi.nlm.nih.gov/nuccore/NC_001665   \n",
       "4        NC_053523   https://www.ncbi.nlm.nih.gov/nuccore/NC_053523.1   \n",
       "\n",
       "   Unnamed: 6  \n",
       "0         NaN  \n",
       "1         NaN  \n",
       "2         NaN  \n",
       "3         NaN  \n",
       "4         NaN  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset = pd.read_csv('species.csv')\n",
    "\n",
    "dataset.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "# TODO: Add 10 more species to your dataset\n",
    "# for species in additional_species:\n",
    "    # Fetch mtDNA and add to the dataset\n",
    "\n",
    "# TODO: Save the expanded dataset\n",
    "# dataset.to_csv('./dataset/expanded_dataset.csv', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "from Bio import Entrez\n",
    "import os\n",
    "from io import BytesIO\n",
    "from requests.adapters import HTTPAdapter\n",
    "from urllib3.util.retry import Retry\n",
    "\n",
    "# Set your email and API key for NCBI\n",
    "entrez_email = os.getenv('moien0813@gmail.com')  \n",
    "entrez_key = os.getenv('f25054f1f99bce38c9f20403bab0f825ef08')\n",
    "Entrez.email = entrez_email \n",
    "\n",
    "def fetch_entrez_record(db, id, rettype, retmode):\n",
    "    \"\"\"Fetch record from NCBI Entrez with retries.\"\"\"\n",
    "    url = f\"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db={db}&id={id}&rettype={rettype}&retmode={retmode}\"\n",
    "    session = requests.Session()\n",
    "    retries = Retry(total=5, backoff_factor=1, status_forcelist=[502, 503, 504])\n",
    "    session.mount('https://', HTTPAdapter(max_retries=retries))\n",
    "\n",
    "    try:\n",
    "        response = session.get(url)\n",
    "        response.raise_for_status()\n",
    "        if db == \"nucleotide\":\n",
    "            return response.text\n",
    "        return response.content\n",
    "    except requests.exceptions.HTTPError as err:\n",
    "        print(f\"HTTP error: {err}\")\n",
    "    except requests.exceptions.ConnectionError as err:\n",
    "        print(f\"Connection error: {err}\")\n",
    "    return None\n",
    "\n",
    "def verify_taxonomy_id(taxo_id, species_name):\n",
    "    \"\"\"Verify taxonomy ID against species name.\"\"\"\n",
    "    xml_data = fetch_entrez_record(\"taxonomy\", taxo_id, \"xml\", \"xml\")\n",
    "    if xml_data:\n",
    "        # Convert bytes data to a binary file-like object\n",
    "        xml_data_io = BytesIO(xml_data)\n",
    "        records = Entrez.read(xml_data_io)\n",
    "        return records[0]['ScientificName'].lower() == species_name.lower()\n",
    "    return False\n",
    "\n",
    "def verify_mitochondrial_dna(accession_number):\n",
    "    gb_data = fetch_entrez_record(\"nucleotide\", accession_number, \"gb\", \"text\")\n",
    "    return \"mitochondrion\" in gb_data.lower() if gb_data else False\n",
    "\n",
    "def extract_accession_from_link(link):\n",
    "    return link.split('/')[-1].split('.')[0]\n",
    "\n",
    "def check_dataset_consistency(file_path):\n",
    "    species_df = pd.read_csv(file_path)\n",
    "\n",
    "    for index, row in species_df.iterrows():\n",
    "        taxonomy_id = str(row['taxo_id'])\n",
    "        species_name = row['specie']\n",
    "        accession_number = row['accession_number']\n",
    "        mtDNA_link = row['mtDNA']\n",
    "        extracted_accession = extract_accession_from_link(mtDNA_link)\n",
    "\n",
    "        taxonomy_check = verify_taxonomy_id(taxonomy_id, species_name)\n",
    "        accession_match = (accession_number == extracted_accession)\n",
    "        mitochondrial_check = verify_mitochondrial_dna(accession_number)\n",
    "\n",
    "        print(f\"Row {index}: Taxonomy Check: {taxonomy_check}, Accession Match: {accession_match}, Mitochondrial Check: {mitochondrial_check}\")\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Row 0: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 1: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 2: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 3: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 4: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 5: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 6: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 7: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 8: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 9: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 10: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 11: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 12: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 13: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 14: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 15: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 16: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 17: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 18: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 19: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 20: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 21: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 22: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 23: Taxonomy Check: False, Accession Match: True, Mitochondrial Check: True\n",
      "Row 24: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 25: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 26: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 27: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 28: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 29: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 30: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 31: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n",
      "Row 32: Taxonomy Check: True, Accession Match: True, Mitochondrial Check: True\n"
     ]
    }
   ],
   "source": [
    "check_dataset_consistency('species.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Entrez, SeqIO\n",
    "import csv\n",
    "\n",
    "# TODO: Write a function to download mtDNA sequences\n",
    "# def download_mtDNA(url, label):\n",
    "    # Code to download and label the sequence\n",
    "def download_fasta(accession, output_file):\n",
    "    Entrez.email = 'moien0813@gmail.com'\n",
    "\n",
    "    handle = Entrez.efetch(db='nucleotide', id=accession, rettype='fasta', retmode='text')\n",
    "    \n",
    "    with open(output_file, 'w') as file:\n",
    "        file.write(handle.read())\n",
    "\n",
    "    handle.close()\n",
    "\n",
    "# TODO: Loop through the dataset and download each mtDNA sequence\n",
    "# for index, row in dataset.iterrows():\n",
    "    # Call the download function for each species\n",
    "csv_file_path = 'species.csv'\n",
    "with open(csv_file_path, 'r') as file:\n",
    "    reader = csv.reader(file)\n",
    "    header = next(reader)\n",
    "    column_index = header.index('accession_number')\n",
    "    accessions = [row[4] for row in reader]\n",
    "\n",
    "\n",
    "# Downloading FASTA files for each accession\n",
    "for accession in accessions:\n",
    "    output_file = f'{accession}.fasta'\n",
    "    download_fasta(accession, output_file)\n",
    "    print(f\"Downloaded FASTA file for accession {accession}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#concatenating downloaded fasta files\n",
    "input_files = []\n",
    "for i in accessions:\n",
    "    input_files.append(i+\".fasta\")\n",
    "\n",
    "\n",
    "output_file = 'concatenated.fasta'\n",
    "\n",
    "with open(output_file, 'w') as out_handle:\n",
    "    for fasta_file in input_files:\n",
    "        with open(fasta_file, 'r') as in_handle:\n",
    "            records = SeqIO.parse(in_handle, 'fasta')\n",
    "        \n",
    "            SeqIO.write(records, out_handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TODO: Install and import the alignment tool\n",
    "# !pip install mafft\n",
    "# from Bio.Align.Applications import MafftCommandline\n",
    "\n",
    "# TODO: Perform the sequence alignment\n",
    "# def perform_alignment(input_file, output_file):\n",
    "    # Code to align sequences using the chosen tool\n",
    "\n",
    "# TODO: Align your downloaded sequences\n",
    "# perform_alignment('path_to_downloaded_sequences.fasta', 'aligned_sequences.fasta')\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from Bio import AlignIO\n",
    "\n",
    "\n",
    "input_file = 'concatenated.fasta'\n",
    "output_file = 'output.fasta'\n",
    "\n",
    "muscle_cline = MuscleCommandline(input=input_file, out=output_file)\n",
    "\n",
    "muscle_cline()\n",
    "\n",
    "alignment = AlignIO.read(output_file, 'fasta')\n",
    "print(alignment)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Conclusion and Reflective Insights\n",
    "\n",
    "As we conclude our exploration of phylogenetic tree construction and analysis, let's reflect on the insights learned from this task and consider questions that emerge from our findings.\n",
    "\n",
    "#### Interpretation of Results:\n",
    "\n",
    "- Reflect on the phylogenetic trees produced by each method (Bayesian inference, maximum likelihood, and neighbor-joining). Consider how the differences in tree topology might offer varied perspectives on the evolutionary relationships among the species.\n",
    "\n",
    "#### Questions to Ponder:\n",
    "\n",
    "1. **Species Divergence**: Based on the trees, which species appear to have the most ancient divergence? How might this information contribute to our understanding of their evolutionary history?\n",
    "   \n",
    "2. **Common Ancestors**: Are there any unexpected pairings or groupings of species that suggest a closer evolutionary relationship than previously thought? How could this reshape our understanding of these species' evolutionary paths?\n",
    "\n",
    "3. **Methodology Insights**: Considering the discrepancies between the trees generated by different methods, what might this tell us about the limitations and strengths of each phylogenetic analysis method?\n",
    "\n",
    "4. **Conservation Implications**: Considering the evolutionary relationships revealed in your phylogenetic analysis, what insights can be gained for conservation strategies? Specifically, how could understanding the close evolutionary ties between species, which might be facing distinct environmental challenges, guide targeted conservation efforts?\n",
    "\n",
    "<blockquote style=\"font-family:Arial; color:red; font-size:16px; border-left:0px solid red; padding: 10px;\">\n",
    "    <strong>Don't forget to answer these questions!</strong>\n",
    "</blockquote>\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
